[MM-18882] Remove ephemeral message for jira server and desktop app

mickmister opened the pull request:
#### Summary

This 4.3.0 desktop release fixes the issue with the Jira Server connect popup [described here](https://mattermost.atlassian.net/browse/MM-18882). This PR enables the popup for users that are running desktop >= v4.3.0

#### Ticket Link

https://mattermost.atlassian.net/browse/MM-18882

#### Additional Info

The wording for the error that shows for desktop < v4.3.0 has changed from:

"Please use your browser to connect to Jira."

to

~~Please upgrade your desktop app to 4.3.0, or log into Mattermost with your browser and run `/jira connect`~~

The current version of your desktop app does not support connecting to Jira directly. Please connect through your browser or [see here](https://github.com/mattermost/mattermost-plugin-jira) for details on upgrading.

mickmister commented:
@DHaussermann I've made it so it only shows the "connect with your browser" message for Jira Server on desktop app < v4.3.0 now

cpoile commented on webapp/src/utils/user_agent.jsx:
```diff
@@ -1,8 +1,35 @@
+    const developDashIndex = userAgent.substring(beginIndex).indexOf('-');
+    const spaceIndex = userAgent.substring(beginIndex).indexOf(' ');
+    const first = Math.min(developDashIndex, spaceIndex);
+
```
1. I know its annoying, but I think we need tests for this -- we're parsing with string indexes here, which is a magnet for errors.
2. Are we sure this works for all the possible desktop app versions?

mickmister replied:
@cpoile Thanks for taking a look at this :+1:

1. I totally agree, not annoying at all

2. I pulled @deanwhillier into this review as he can shed light on the nuances of the desktop app. The code here certainly seems rickety at the moment.

deanwhillier replied:
I took a quick look at the Desktop source and it doesn't look like we're setting the agent explicitly, which means Electron is doing so on our behalf. I think it's safe to say that it is consistent. That said, relying on the user agent might not be the best approach? 🤷‍♂️

For v4.3.0, I implemented a rudimentary communication bridge between the Desktop App and the Webapp to handle user activity updates from the Desktop app as well as for the Webapp to send desktop notification requests to the Desktop App for handling (because, many reasons). In doing this I added a mechanism for the Desktop app to register itself with the Webapp once a user has logged in, including passing its version number. That version is currently stored as a string in `window.desktop.version` if it is a valid semantic version (not the best spot long term). I added and am using the `semver` package to parse, validate and compare version numbers: https://www.npmjs.com/package/semver.

So, as of 5.16 (backported to lastest 5.9, 5.14 & 5.15), if the Desktop App v4.3.0+ loads the Webapp, the Webapp will have access to the Desktop Version code it is running in via `window.desktop.version` once the user has logged in and you can use `semver` to do comparisons as needed.

mickmister replied:
@deanwhillier Your suggestion seems like a better approach because it is more within our control. I'm wondering if the error we show here will suffice (considering the webapp version also matters here). I think making the error as user-friendly as possible is the way to combat this.

deanwhillier replied:
@mickmister, I guess that depends on which server/webapp version(s) the Jira plugin can run on? If it can run on servers that don't include the `window.desktop.version` property, then the ephemeral message would probably need to mention a server version in addition to the desktop app version needed for connecting in-app. Challenge is wording the message so as not to be too technical or verbose.

Maybe something like "The current version of your desktop app doesn't support connecting to Jira directly, please connect through your browser or [see here](tbd) for details on upgrading." The link could point to our docs with a more detailed description of what is needed to connect to Jira with the desktop app including version numbers of both the app and server. It's also easier to managed details like these in a doc vs in the plugin itself. Just a thought! Might want to run this by the PM?

mickmister replied:
Excellent thinking! We can offload these complex explanations to the docs. @aaronrothschild What are your thoughts on this?

mickmister replied:
@aaronrothschild Ping for this, I want to get your input on this before I move forward with next steps

aaronrothschild commented:
@mickmister Sorry, I forget if we discussed this part: What should the user DO?

Can we change it to "Please use your regular browser to login to Mattermost - then use the `/jira connect` command to connect your Jira login with Mattermost."?

mickmister commented:
I've updated the PR to use `window.desktop.version` to detect the desktop app version. I've also changed the error message to include a link to the docs.

Here's a ticket for docs needed for Jira 2.3: https://mattermost.atlassian.net/browse/MM-19933

cc @aaronrothschild

mickmister commented on webapp/src/hooks/hooks.js:
```diff
@@ -35,8 +35,10 @@ export default class Hooks {
-                this.store.dispatch(sendEphemeralPost('Please use your browser to connect to Jira.'));
+
+            if (getInstalledInstanceType(this.store.getState()) === 'server' && isDesktopApp() && !isMinimumDesktopAppVersion(4, 3, 0)) { // eslint-disable-line no-magic-numbers
+                const errMsg = 'The current version of your desktop app does not support connecting to Jira directly. Please connect through your browser or [see here](https://github.com/mattermost/mattermost-plugin-jira) for details on upgrading.';
```
Here is the error message we show for when the Jira Server user has desktop app before `v4.3.0`

aaronrothschild commented on webapp/src/hooks/hooks.js:
```diff
@@ -35,8 +35,10 @@ export default class Hooks {
-                this.store.dispatch(sendEphemeralPost('Please use your browser to connect to Jira.'));
+
+            if (getInstalledInstanceType(this.store.getState()) === 'server' && isDesktopApp() && !isMinimumDesktopAppVersion(4, 3, 0)) { // eslint-disable-line no-magic-numbers
+                const errMsg = 'The current version of your desktop app does not support connecting to Jira directly. Please connect through your browser or [see here](https://github.com/mattermost/mattermost-plugin-jira) for details on upgrading.';
```
```suggestion
                const errMsg = 'Your version of the Mattermost Desktop Client does not support authenticating between Jira and Mattermost directly.  To connect your Jira account with Mattermost, Go to Mattermost via your web browser and then type `/jira connect` or [check the Mattermost download page](https://mattermost.com/download/#mattermostApps) to get the latest version of the desktop client';
```
@mickmister We want them to update their desktop client not the jira plugin (unless I'm missing something).

aaronrothschild replied:
@mickmister is there any way for us to make a link like `[GO HERE](https://<MM_SITEURL>)` so they can just click on the link, instead of them trying to find out where their mattermost server is located?

mickmister replied:
Unfortunately the desktop app opens up `https://<MM_SITEURL>` in a popup since it is the same domain, so it doesn't open the browser. We can display it with `` so it can be used for copy and paste, and not render as a clickable link.

mickmister replied:
@aaronrothschild The original wording of the error was pointing to Jira docs, because I was thinking we would have a full explanation of what's needed to upgrade, including having server `5.16` or other backport-supported server versions. If we can assume they have an up-to-date server version, then we can link to the desktop app download page like you suggest. That seems most appropriate for us too.

mickmister replied:
@aaronrothschild Do you mind taking another look at this when you have a chance?

DHaussermann commented:
@mickmister I saw an issue the first couple times after authenticating and clicking "Allow", it would say "request token mismatch"
![DesktopMapping](https://user-images.githubusercontent.com/35588783/69100961-56d72300-0a2c-11ea-864b-ddb920e1cf9f.PNG)
Any ideas as to what may have been happening? (only occurred on Server instance)

mickmister commented:
@DHaussermann This is an error generated by the plugin itself. I'm not sure exactly what causes it.

If I understand correctly, we:
- Generate an oauth token
- Send it to Jira Server
- Assert that after the user approves the connection that the token we initially sent Jira is equal to the one that Jira says we sent

Here's the code for reference

https://github.com/mattermost/mattermost-plugin-jira/blob/13d121f8e6791eb2f9c35b5e181f659c613e3189/server/instance_server.go#L74-L83

https://github.com/mattermost/mattermost-plugin-jira/blob/13d121f8e6791eb2f9c35b5e181f659c613e3189/server/user_server.go#L60-L67

The assertion on line 65 is failing, meaning the token we receive back from Jira is not equal to the one we sent. @levb Do you have an example cause of this situation?

DHaussermann commented:
Token mismatch issue aside, This is working...
- Desktop 4.2 w\ Cloud --> Map User
- Desktop 4.2 w\ server --> **New ephemeral message is shown**

- Desktop 4.3 w\ Cloud --> Map User
- Desktop 4.3 w\ server --> Map User

- Chrome w\ cloud --> Map User
- Chrome w\ server --> Map User

Regarding the intermittent issue, this seems to occur for me whenever the desktop app's session is not yest authenticated on the Jira server. Once I have provided credentials once and the window opens directly to the mapping page, I can no longer reproduce the issue. 
@levb or @mickmister  does this help narrow down this issue?

DHaussermann commented:
Spoke with @levb and we agreed that we should address the token mismatch as a separate task. 

Created https://mattermost.atlassian.net/browse/MM-20346 to address this. I will update the task if I am able to find repro steps.
